// react-preview-file/dist/filePreview.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = require("react");
var FilePreview = (function (_super) {
    __extends(FilePreview, _super);
    function FilePreview() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FilePreview.prototype.componentWillReceiveProps = function (_a) {
        var newFile = _a.file;
        var file = this.props.file;
        if (file !== newFile) {
            this.revoke();
            this.createPreview(newFile);
        }
    };
    FilePreview.prototype.componentWillMount = function () {
        var file = this.props.file;
        this.createPreview(file);
    };
    FilePreview.prototype.componentWillUnmount = function () {
        this.revoke();
    };
    FilePreview.prototype.createPreview = function (file) {
        this.preview = URL.createObjectURL(file);
    };
    FilePreview.prototype.revoke = function () {
        URL.revokeObjectURL(this.preview);
    };
    FilePreview.prototype.render = function () {
        var children = this.props.children;
        return children(this.preview);
    };
    return FilePreview;
}(react_1.Component));
exports.FilePreview = FilePreview;
exports.default = FilePreview;
